src/sys/pthread.rs

```diff
@@ -34,10 +41,285 @@ pub fn pthread_kill<T>(thread: Pthread, signal: T) -> Result<()>
+            Errno::result(libc::pthread_mutexattr_getpshared(&self.0,uninit.as_mut_ptr()))?;
+            uninit.assume_init()
+        };
+        Ok(init == libc::PTHREAD_PROCESS_SHARED)
+    }
+    /// Wraps [`libc::pthread_mutexattr_setpshared`].
+    pub fn set_shared(&mut self, shared: bool) -> Result<()> {
+        let shared = if shared { libc::PTHREAD_PROCESS_SHARED} else { libc::PTHREAD_PROCESS_PRIVATE };
+        unsafe {
+            Errno::result(libc::pthread_mutexattr_setpshared(&mut self.0,shared)).map(drop)
+        }
+    }
+    /// Wraps [`libc::pthread_mutexattr_getrobust`].
+    pub fn get_robust(&self) -> Result<bool> {
+        let init = unsafe {
+            let mut uninit = std::mem::MaybeUninit::uninit();
+            Errno::result(libc::pthread_mutexattr_getrobust(&self.0,uninit.as_mut_ptr()))?;
+            uninit.assume_init()
+        };
+        Ok(init == libc::PTHREAD_MUTEX_ROBUST)
+    }
+    /// Wraps [`libc::pthread_mutexattr_setrobust`].
+    pub fn set_robust(&mut self, robust: bool) -> Result<()> {
+        let robust = if robust { libc::PTHREAD_MUTEX_ROBUST} else { libc::PTHREAD_MUTEX_STALLED };
+        unsafe {
+            Errno::result(libc::pthread_mutexattr_setrobust(&mut self.0,robust)).map(drop)
+        }
+    }
+    /// Wraps [`libc::pthread_mutexattr_getprotocol`].
+    pub fn get_protocol(&self) -> Result<Protocol> {
+        let init = unsafe {
+            let mut uninit = std::mem::MaybeUninit::uninit();
+            Errno::result(libc::pthread_mutexattr_getprotocol(&self.0,uninit.as_mut_ptr()))?;
+            uninit.assume_init()
+        };
+        Ok(Protocol::from(init))
+    }
+    /// Wraps [`libc::pthread_mutexattr_setprotocol`].
+    pub fn set_protocol(&mut self, protocol: Protocol) -> Result<()> {
+        unsafe {
+            Errno::result(libc::pthread_mutexattr_setprotocol(&mut self.0,protocol as i32)).map(drop)
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+impl Default for MutexAttr {
+    fn default() -> Self {
+        let mutex_attr = Self::new().unwrap();
+        debug_assert_eq!(mutex_attr.get_shared(),Ok(true));
+        mutex_attr
+    }
+}
+#[cfg(target_os = "linux")]
+impl std::ops::Drop for MutexAttr {
+    /// Wraps [`libc::pthread_mutexattr_destroy`].
+    fn drop(&mut self) {
+        unsafe {
+            Errno::result(libc::pthread_mutexattr_destroy(&mut self.0)).unwrap();
+        }
+    }
+}
+
+/// Mutex.
+/// ```
+/// # use std::{
+/// #   sync::Arc,
+/// #   time::{Instant, Duration},
+/// #   thread::{sleep, spawn},
+/// #   mem::size_of,
+/// #   num::NonZeroUsize,
+/// #   os::unix::io::OwnedFd
+/// # };
+/// # use nix::{
+/// #   sys::{pthread::{Mutex, MutexAttr}, mman::{mmap, MapFlags, ProtFlags}},
+/// #   unistd::{fork,ForkResult},
+/// # };
+/// const TIMEOUT: Duration = Duration::from_millis(500);
+/// const DELTA: Duration = Duration::from_millis(100);
+/// # fn main() -> nix::Result<()> {
```

Comment:
```suggestion
```
It's normally not necessary to put an explicit `main` function in a doc test.

Reply:
I believe it is necessary so as to use a specific error type.

Otherwise it automatically defines it `main -> ()` which doesn't work here.

Reply:
Oh, ok.  Usually I just `.unwrap()` in doc tests, so return type isn't an issue.

Reply:
Yeah, its not important, its more just ✨ aesthetic ✨.